components/ftx/browser/ftx_service.cc

```diff
@@ -0,0 +1,398 @@
+
+namespace {
+
+const char api_host[] = "ftx.com";
+const char oauth_callback[] = "com.brave.ftx://authorization";
+const unsigned int kRetriesCountOnNetworkChange = 1;
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("ftx_service", R"(
+      semantics {
+        sender: "FTX Service"
+        description:
+          "This service is used to communicate with FTX "
+          "on behalf of the user interacting with the FTX widget."
+        trigger:
+          "Triggered by using the FTX widget."
+        data:
+          "Account balance for the widget."
+        destination: WEBSITE
+      }
+      policy {
+        cookies_allowed: NO
+        setting:
+          "You can enable or disable this feature on the new tab page."
+        policy_exception_justification:
+          "Not implemented."
+      }
+    )");
+}
+
+GURL GetURLWithPath(const std::string& host, const std::string& path) {
+  return GURL(std::string(url::kHttpsScheme) + "://" + host).Resolve(path);
+}
+
+void BuildFormEncoding(const std::string& key,
+                       const std::string& value,
+                       std::string* out) {
+  if (!out->empty())
+    out->append("&");
+  out->append(key + "=" + net::EscapeUrlEncodedData(value, true));
+}
+
+}  // namespace
+
+FTXService::FTXService(content::BrowserContext* context)
+    : client_id_(FTX_CLIENT_ID),
+      client_secret_(FTX_CLIENT_SECRET),
+      context_(context),
+      url_loader_factory_(
+          content::BrowserContext::GetDefaultStoragePartition(context_)
+              ->GetURLLoaderFactoryForBrowserProcess()),
+      weak_factory_(this) {
+  PrefService* prefs = user_prefs::UserPrefs::Get(context);
+  // Get access token from prefs
+  std::string encoded_encrypted_access_token =
+      prefs->GetString(kFTXAccessToken);
+  std::string encrypted_access_token;
+  if (!base::Base64Decode(encoded_encrypted_access_token,
+                          &encrypted_access_token)) {
+    LOG(ERROR) << "FTX: Could not decode Token info from prefs";
+  } else {
+    if (!OSCrypt::DecryptString(encrypted_access_token, &access_token_)) {
+      LOG(ERROR) << "FTX: Could not decrypt and save access token";
+    }
+  }
+}
+
+FTXService::~FTXService() {}
+
+bool FTXService::GetFuturesData(GetFuturesDataCallback callback) {
+  auto internal_callback = base::BindOnce(
+      &FTXService::OnFuturesData, base::Unretained(this), std::move(callback));
+  GURL url = GetURLWithPath(api_host, get_futures_data_path);
+  return NetworkRequest(url, "GET", "", "",
+      std::move(internal_callback), false);
+}
+
+void FTXService::OnFuturesData(
+    GetFuturesDataCallback callback,
+    const int status,
+    const std::string& body,
+    const std::map<std::string, std::string>& headers) {
+  FTXFuturesData data;
+  if (status >= 200 && status <= 299) {
+    FTXJSONParser::GetFuturesDataFromJSON(body, &data, futures_filter);
```

Comment:
return value is ignored

Reply:
Empty data will be returned. If we handle `false` return value then we can show error UI. Follow-up here https://github.com/brave/brave-browser/issues/15852